src/base/Flow.ts

```diff
@@ -0,0 +1,314 @@
+import { v4 as uuidV4 } from "uuid";
+import { RedisClientOptions, createClient } from "redis";
+import Queue from "./Queue";
+import Worker from "./Worker";
+import {
+  AnyObject,
+  FlowOptions,
+  ILogger,
+  RedisClient,
+  WorkerFunction,
+  IStartable,
+  BaseJob,
+  ArrayElement,
+} from "./types";
+import { objectToStringEntries, parseObject } from "../utils";
+import ParentWorker from "./ParentWorker";
+import { JOB_KEY_PREFIX } from "../static";
+
+/**
+ * Defines a sequence of Jobs / Queues / Workers
+ */
+export default class Flow<
+  FlowJob extends BaseJob,
+  CreateJobOptions extends AnyObject,
+  LookupAttributes extends keyof CreateJobOptions
+> {
+  /**
+   * Name of the flow
+   */
+  public readonly name: string;
+
+  /**
+   * Provided logger
+   */
+  private readonly logger: ILogger;
+
+  /**
+   * Attributes which can be used for lookup a job
+   */
+  private readonly lookupAttributes: string[];
+
+  /**
+   * Queues of the Flow
+   */
+  private readonly queues: Queue[];
+
+  /**
+   * Workers of the Flow
+   */
+  private workers: IStartable[] = [];
+
+  /**
+   * Options to create redis connections
+   */
+  private readonly redisClientOptions: RedisClientOptions;
+
+  /**
+   * Redis client instance
+   */
+  private readonly redis: RedisClient;
+
+  /**
+   * Set the basic options, initialize queues and redis client
+   * @param options parameters of AccessFlow
+   */
+  constructor(options: FlowOptions) {
+    const { name, logger, redisClientOptions, queueOptions, lookupAttributes } =
+      options;
+
+    this.name = name;
+    this.logger = logger;
+    this.lookupAttributes = lookupAttributes;
+    this.redisClientOptions = redisClientOptions;
+    this.redis = createClient(redisClientOptions);
```

Comment:
Consider implementing dependency injection here.

Reply:
The reason I did this, is because we need `(num of Flows) + (num of Workers) * 2` redis client instances, so injecting all of them would be painful.

Reply:
I’m pretty sure there are design patterns that allow such things